app/agent/workflows/shared-workflow/index.ts

```diff
@@ -0,0 +1,161 @@
+/* eslint-disable no-console */
+import assert from "assert";
+
+import {
+  condition,
+  getExternalWorkflowHandle,
+  setHandler,
+  startChild,
+  ParentClosePolicy,
+  Trigger,
+  proxyActivities,
+  uuid4,
+  isCancellation,
+  defineSignal,
+  CancellationScope,
+} from "@temporalio/workflow";
+
+import type { Activities } from "~/agent/activities/list";
+import type { WaitRequest } from "~/agent/activities/shared-workflow/shared";
+import { sharedWorkflowIdQuery } from "~/agent/activities/shared-workflow/shared";
+import { requestsQuery } from "~/agent/activities/shared-workflow/shared";
+import { WorkflowCancelledError } from "~/agent/activities/shared-workflow/shared";
+import { WaitRequestType } from "~/agent/activities/shared-workflow/shared";
+import { waitRequestSignal } from "~/agent/activities/shared-workflow/shared";
+import { retrySignal } from "~/agent/workflows-utils";
+import { wrapWorkflowError } from "~/temporal/utils";
+
+type GetWorkflowResult =
+  | { ok: true; result: unknown; error?: undefined }
+  | { ok: false; result?: undefined; error: unknown };
+
+const { signalWithStartWaitWorkflow } = proxyActivities<Activities>({
+  startToCloseTimeout: "30 seconds",
+  retry: {
+    maximumAttempts: 10,
+  },
+});
+
+export async function runSharedWorkflow(args: {
+  workflow: string;
+  params: unknown[];
+}): Promise<void> {
+  const requests: WaitRequest[] = [];
+  const workflowId = uuid4();
+  const cancelledTrigger = new Trigger<"cancelled">();
+  let cancelled = 0;
+  setHandler(waitRequestSignal, (req: WaitRequest) => {
+    const idx = requests.findIndex((e) => e.releaseSignalId === req.releaseSignalId);
+    if (idx >= 0 && requests[idx].releaseSignalId === WaitRequestType.CANCEL) {
+      return;
+    }
+    if (req.type === WaitRequestType.CANCEL) {
+      cancelled++;
+    }
+    if (idx >= 0) {
+      requests[idx] = req;
+    } else {
+      requests.push(req);
+    }
+    if (cancelled >= requests.length) {
+      cancelledTrigger.resolve("cancelled");
+    }
+  });
+  setHandler(requestsQuery, () => requests);
+  setHandler(sharedWorkflowIdQuery, () => workflowId);
+  let result: GetWorkflowResult = {
+    ok: false,
+    error: new Error("Unknown shared workflow error"),
+  };
+  try {
+    const handle = await startChild(args.workflow, {
+      args: args.params as any,
+      parentClosePolicy: ParentClosePolicy.PARENT_CLOSE_POLICY_ABANDON,
+      workflowId,
+    });
+    const partialResult = await Promise.race([
+      cancelledTrigger,
+      handle
+        .result()
+        .then((result) => ({
+          ok: true as const,
+          result,
+        }))
+        .catch((error) => ({
+          ok: false as const,
+          error,
+        })),
+    ]);
+    if (partialResult === "cancelled") {
+      const externalHandle = getExternalWorkflowHandle(workflowId);
+      await externalHandle.cancel().catch((err) => {
```

Comment:
Perhaps you should await the `handle.result()` promise here so it's not dangling after exiting this funtion

Reply:
I don't want to do it. This dangling promise can cause no harm - both resolve and reject paths are covered. Awaiting it could cause the runSharedWorklow to hang if the workflow being cancelled takes a long time to do so, for example if it just called an activity that's not cancellable.